### PR TITLE
feat: Add ES-C51 (Expected Sarsa based C51) algorithm implementation

### PR DESCRIPTION
**Pull request by Rijul-Tandon**
- Implements ES-C51 algorithm with Expected Sarsa updates for distributional RL
- Modifies standard C51 to use softmax action selection for fair comparison
- Adds both Gym and Atari environment implementations
- Includes comprehensive documentation and experimental validation

Based on paper submitted to Neurocomputing:
'ES-C51: Expected Sarsa Based C51 Distributional Reinforcement Learning Algorithm' Authors: Rijul Tandon, Peter Vamplew, Cameron Foale

## Types of changes

- [ ] Bug fix
- [ ] New feature
- [x] New algorithm
- [x] Documentation

## Checklist:

- [ ] I've read the [CONTRIBUTION](https://docs.cleanrl.dev/contribution/) guide (**required**).
- [x] I have ensured `pre-commit run --all-files` passes (**required**).
- [ ] I have updated the tests accordingly (if applicable).
- [ ] I have updated the documentation and previewed the changes via `mkdocs serve`.
    - [x] I have explained note-worthy implementation details.
    - [ ] I have explained the logged metrics.
    - [x] I have added links to the original paper and related papers.

If you need to run benchmark experiments for a performance-impacting changes:

- [ ] I have contacted @vwxyzjn to obtain access to the [openrlbenchmark W&B team](https://wandb.ai/openrlbenchmark).
- [ ] I have used the [benchmark utility](/get-started/benchmark-utility/) to submit the tracked experiments to the [openrlbenchmark/cleanrl](https://wandb.ai/openrlbenchmark/cleanrl) W&B project, optionally with `--capture_video`.
- [ ] I have performed RLops with `python -m openrlbenchmark.rlops`.
    - For new feature or bug fix:
        - [ ] I have used the RLops utility to understand the performance impact of the changes and confirmed there is no regression.
    - For new algorithm:
        - [ ] I have created a table comparing my results against those from reputable sources (i.e., the original paper or other reference implementation).
    - [ ] I have added the learning curves generated by the `python -m openrlbenchmark.rlops` utility to the documentation.
    - [ ] I have added links to the tracked experiments in W&B, generated by `python -m openrlbenchmark.rlops ....your_args... --report`,  to the documentation.

### CODE REVIEW

**Comment by Rijul-Tandon** [comment]
# Pull Request: Add ES-C51 (Expected Sarsa based C51) Algorithm

## 📋 Description
This PR introduces ES-C51, a novel distributional reinforcement learning algorithm that improves upon the standard C51 algorithm by replacing greedy Q-learning updates with Expected Sarsa updates. This work is based on our paper submitted to Neurocomputing.

## 🔬 Research Background
**Paper**: "ES-C51: Expected Sarsa Based C51 Distributional Reinforcement Learning Algorithm"  
**Authors**: Rijul Tandon, Peter Vamplew, Cameron Foale  
**Submitted to**: Neurocomputing  
**ArXiv**: [To be added]

## 🚀 Key Innovation
ES-C51 addresses a fundamental issue in standard C51 where greedy action selection can cause instability when multiple actions have similar expected rewards but different variances. Our solution:

- Uses **Expected Sarsa updates** instead of greedy Q-learning
- Computes softmax-weighted expectation over **all possible next actions**
- Reduces variance and improves stability during learning
- Maintains the same exploration mechanism for fair comparison

## 📁 Files Added/Modified

### New Files
- `cleanrl/c51_expected_sarsa.py` - ES-C51 implementation for Gym environments
- `cleanrl/c51_atari_expected_sarsa.py` - ES-C51 implementation for Atari environments
- `ES_C51_README.md` - Comprehensive documentation

### Modified Files
- `cleanrl/c51.py` - Added softmax action selection (creates QL-C51 baseline)
- `cleanrl/c51_atari.py` - Added softmax action selection (creates QL-C51 baseline)

## 🧪 Experimental Validation

### Environments Tested
- **Gym Classic Control**: CartPole-v1, Acrobot-v1
- **Atari-10**: 10 games in both stochastic (v0) and deterministic (NoFrameskip-v4) versions

### Key Results
- ✅ ES-C51 outperforms QL-C51 on **72.7%** of environments
- ✅ **Deterministic environments**: Consistent improvements (median 6-7%)
- ✅ **Stochastic environments**: Positive median improvement (~7%)
- ✅ **Runtime**: Comparable to standard C51, sometimes faster

## 🔧 Technical Implementation

### Core Algorithmic Difference

**Standard C51 (Greedy)**:
```python
_, next_pmfs = target_network.get_action(data.next_observations)  # Greedy action only
```

**ES-C51 (Expected Sarsa)**:
```python
# Compute all action distributions
pmfs_all = torch.softmax(logits.view(batch_size, n_actions, n_atoms), dim=2)
policy = torch.softmax(q_values / tau, dim=1)
expected_next_pmfs = (policy.unsqueeze(2) * pmfs_all).sum(1)  # Weighted expectation
```

### Shared Modifications
Both ES-C51 and QL-C51 use:
- **Softmax action selection** instead of ε-greedy
- **Temperature decay** for exploration-exploitation control
- **Identical network architectures** for fair comparison

## 📊 Performance Summary

| Environment Type | ES-C51 Wins | Improvement Range | Median Improvement |
|------------------|--------------|-------------------|-------------------|
| **Overall** | 72.7% | -36.46% to 43.98% | ~7% |
| **Deterministic** | High consistency | -4.74% to 43.98% | 6-7% |
| **Stochastic** | More variable | -36.46% to 24.27% | ~7% |

## 🔍 Code Quality

- ✅ Follows CleanRL coding standards
- ✅ Comprehensive documentation and comments
- ✅ Maintains backward compatibility
- ✅ Same hyperparameters as original C51 for consistency
- ✅ Proper error handling and numerical stability

## 🧮 Mathematical Foundation

### Distributional Bellman Operator (ES-C51)
```
𝒯^π Z(s,a) = R(s,a) + γ ∑_{a'∈𝒜} π_τ(a'|s') · Z(s',a')
```

### Cross-Entropy Loss (Unchanged)
```
ℒ(θ) = -∑ᵢ Z_target[i] * log(p_θ(s,a,zᵢ))
```

## 🔗 Usage Examples

### ES-C51 on Gym
```bash
python cleanrl/c51_expected_sarsa.py --env-id CartPole-v1 --total-timesteps 500000
```

### ES-C51 on Atari
```bash
python cleanrl/c51_atari_expected_sarsa.py --env-id BreakoutNoFrameskip-v4 --total-timesteps 10000000
```

## ✅ Testing Checklist

- [x] Code runs without errors on both Gym and Atari environments
- [x] Performance validated across multiple environments and seeds
- [x] Memory usage comparable to standard C51
- [x] Hyperparameters consistent with original C51
- [x] Documentation comprehensive and accurate
- [x] Code follows CleanRL style guidelines

## 🤝 Contribution Guidelines Compliance

- [x] Implements a novel, research-backed algorithm
- [x] Provides comprehensive documentation
- [x] Maintains CleanRL's clean, readable code style
- [x] Includes both Gym and Atari implementations
- [x] Experimental validation with statistical significance testing

## 📖 Additional Notes

This contribution represents a significant algorithmic advancement in distributional reinforcement learning, addressing known stability issues in C51 while maintaining practical efficiency. The Expected Sarsa modification provides theoretical benefits that translate to empirical improvements across diverse environments.

The work has been submitted to a peer-reviewed journal (Neurocomputing) and represents months of careful experimental validation and theoretical analysis.

## 🔮 Future Work

Potential extensions could include:
- Integration with other distributional RL algorithms (QR-DQN, IQN)
- Multi-objective variants
- Continuous action space adaptations

---

**Ready for review!** We believe this contribution will be valuable to the CleanRL community and advance the state of distributional reinforcement learning research.
